### PR TITLE
Fix issue where explicitly specified targets would be passed to the separate restore command

### LINKED ISSUE #15485: dotnet cli passing -target:Restore to MSBuild instead of -restore
As seen in: https://github.com/xamarin/net6-samples/pull/52

We have a [multi-targeted](https://github.com/xamarin/net6-samples/blob/e0259e3d94f12a9184623905d9e07c1df8567808/HelloForms/HelloForms.csproj#L3) project with:
```xml
<TargetFrameworks>net6.0-android;net6.0-ios</TargetFrameworks>
```

To deploy to an Android device/emulator (and _not_ iOS), you can run:
```cmd
> dotnet build -t:Run -p:TargetFramework=net6.0-android
```

With .NET 6.0.100-alpha.1.21064.27, this command fails with:
```
error MSB4057: The target "Run" does not exist in the project.
```

The MSBuild arguments the dotnet cli is passing seem odd:
```
-target:Restore -t:Run
```

With .NET 6.0.100-alpha.1.20562.2, it used to pass:
```
-restore -t:Run
```

In fact, you can workaround the problem by doing:
```cmd
> dotnet build -t:Run -p:TargetFramework=net6.0-android --no-restore
```

I tried creating other, simpler projects to reproduce this issue, but a console app with `TargetFrameworks=net5.0;net6.0` works fine. I always see `-restore` being passed.

I think it is something to with .NET workloads and multi-targeting at the same time?

* Example broken log: [broken.zip](https://github.com/dotnet/sdk/files/5869975/broken.zip)
* Example working log: [works.zip](https://github.com/dotnet/sdk/files/5869960/works.zip)

### PR DESCRIPTION
**Pull request by sfoslund**
Fixes https://github.com/dotnet/sdk/issues/15485

### CODE REVIEW

File: src/Cli/dotnet/commands/RestoringCommand.cs [hunk 1]
```diff
@@ -73,14 +65,19 @@ private static IEnumerable<string> Prepend(string argument, IEnumerable<string>
...
 
         //  These arguments don't by themselves require that restore be run in a separate process,
         //  but if there is a separate restore process they shouldn't be passed to it
         private static bool IsExcludedFromSeparateRestore(string argument)
-            => argument.StartsWith("-t:", StringComparison.Ordinal) ||
-               argument.StartsWith("-target:", StringComparison.Ordinal);
+            => propertyPrefixes.Any(prefix => argument.StartsWith($"{prefix}t:", StringComparison.Ordinal)) ||
+               propertyPrefixes.Any(prefix => argument.StartsWith($"{prefix}target:", StringComparison.Ordinal)) ||
+               propertyPrefixes.Any(prefix => argument.StartsWith($"{prefix}consoleloggerparameters:", StringComparison.Ordinal)) ||
+               propertyPrefixes.Any(prefix => argument.StartsWith($"{prefix}clp:", StringComparison.Ordinal));
```
Comment (dsplaisted) [inline src/Cli/dotnet/commands/RestoringCommand.cs:80]: Why don't we want to pass console logger parameters to the restore command?

File: src/Tests/dotnet.Tests/dotnet-msbuild/GivenDotnetBuildInvocation.cs [hunk 2]
```diff
@@ -54,8 +54,10 @@ public void MsbuildInvocationIsCorrect(string[] args, string expectedAdditionalA
 
         [Theory]
         [InlineData(new string[] { "-f", "tfm" }, "-target:Restore", "-property:TargetFramework=tfm")]
-        [InlineData(new string[] { "-p:TargetFramework=tfm" }, "-target:Restore", "-property:TargetFramework=tfm")]
+        [InlineData(new string[] { "-p:TargetFramework=tfm" }, "-target:Restore", "-p:TargetFramework=tfm")]
+        [InlineData(new string[] { "/p:TargetFramework=tfm" }, "-target:Restore", "-property:TargetFramework=tfm")]
```
Comment (dsplaisted) [inline src/Tests/dotnet.Tests/dotnet-msbuild/GivenDotnetBuildInvocation.cs:58]: What's the reason that `/p` gets translated to `-property` but `-p` stays as is?

**Review by dsplaisted — APPROVED** [review]
Do you know what the intent behind having separate msbuildArgs, parsedArguments, and trailingArguments was?  I don't understand it at all so I'm a bit worried that removing them could break something.

**Comment by dsplaisted** [comment]
FYI @Redth @jonathanpeppers

File: src/Cli/dotnet/commands/RestoringCommand.cs (hunk 1 above)
Comment (sfoslund) [reply src/Cli/dotnet/commands/RestoringCommand.cs:80]: When we include it we get output for the separate restore as well as the main target, so it results in having 2 separate "0 errors, build succeeded" messages

**Comment by sfoslund** [comment]
> Do you know what the intent behind having separate msbuildArgs, parsedArguments, and trailingArguments was? I don't understand it at all so I'm a bit worried that removing them could break something.

I'm not totally sure either, @wli3 do you know the history?

File: src/Tests/dotnet.Tests/dotnet-msbuild/GivenDotnetBuildInvocation.cs (hunk 2 above)
Comment (sfoslund) [reply src/Tests/dotnet.Tests/dotnet-msbuild/GivenDotnetBuildInvocation.cs:58]: `/p` had to be hardcoded as a option to work around a system command line issue, but `-p` works fine by just being parsed/forwarded as is https://github.com/dotnet/sdk/blob/master/src/Cli/dotnet/CommonOptions.cs#L15

**Comment by wli3** [comment]
sorry I don't know either

**Comment by wli3** [comment]
@jonsequitur ?

**Comment by Redth** [comment]
Would be great if this can make p3 🙂  as it’s causing some confusion in the single project .NET MAUI experience.

**Comment by dsplaisted** [comment]
I'm still interested in understanding what was going on with all those different types of arguments, but let's not block on that.

**Pull request merged**